src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/ModelStateDictionary.cs

```diff
@@ -521,14 +523,41 @@ private ModelStateNode GetNode(string key, bool createIfNotExists)
             //    -> baz
             //     -> [qux]
 
+            DelimiterMatch lastMatchType = DelimiterMatch.NoMatch;
+
             var current = _root;
             var previousIndex = 0;
-            int index;
-            while ((index = key.IndexOfAny(Delimiters, previousIndex)) != -1)
+            int index = 0;
+
+            foreach (var ch in key)
             {
-                var keyStart = previousIndex == 0 || key[previousIndex - 1] == '.'
-                    ? previousIndex
-                    : previousIndex - 1;
+                DelimiterMatch matchType;
+                switch (ch)
```

Comment:
Would it be easier to read if this was refactored in to a separate method?

```
var matchResult = FindNext(key, index);
matchResult.MatchType, matchResult.Index ..
```

Not that it's particularly difficult to read now.